### PR TITLE
move import match to base

### LINKED ISSUE #5: Extend Importing Function with the Import Matching
https://github.com/OpenSPP/openspp-registry/pull/237#pullrequestreview-1492306958

move the code to openspp_base

How to Use:
1. Create an Import Matching

- Go to Registry > Configuration > Import Match
- Click Create Button
- Select the Model for the Import Match
- Add the Fields to be used in Import Matching
- Check Overwrite Match if you want Matched Data to be overwritten or don't check if you want to Ignore those matched data.
- Click Save.

### PR DESCRIPTION
**Pull request by emjay0921**

### CODE REVIEW

**Review by gonzalesedwin1123 — APPROVED** [review]

File: spp_import_match/models/import_match.py [hunk 1]
```diff
@@ -1,126 +1,149 @@
...
+                    fields_list.append(field.field_id.id)
+
+            duplicate_counter = 0
+            for duplicate_field in fields_list:
+                if duplicate_field == field_id:
+                    duplicate_counter += 1
+
+            if duplicate_counter > 1:
+                raise ValidationError(_("Field '%s', already exists!") % rec.field_id.field_description)
+
```
Comment (nhatnm0612) [inline spp_import_match/models/import_match.py]:
```suggestion
_sql_constraints = [
    (
        "field_id_uniq",
        "UNIQUE(field_id, match_id)",
        "Field should be unique for each match",
    )
]


```

File: spp_import_match/models/import_match.py [hunk 2]
```diff
@@ -1,126 +1,149 @@
...
+    model_name = fields.Char(related="model_id.model")
+    model_description = fields.Char(related="model_id.name")
+    field_ids = fields.One2many(
+        "spp.import.match.fields",
+        "match_id",
+        string="Fields",
+        required=True,
+        help="Fields to Match in Importing",
+    )
+
```
Comment (nhatnm0612) [inline spp_import_match/models/import_match.py:35]:
```suggestion

_sql_constraints = [
    (
        "model_id_uniq",
        "UNIQUE(model_id)",
        "Model should be unique!",
    )
]
```

File: spp_import_match/models/import_match.py [hunk 3]
```diff
@@ -1,126 +1,149 @@
...
+    @tools.ormcache("model_name", "frozenset(fields)")
+    def _usable_rules(self, model_name, fields):
+        result = self
+        available = self.search([("model_name", "=", model_name)])
+        field_to_match = []
+        for record in available:
+            field_to_match.append(record.field_ids.mapped("name"))
+            for f in record.field_ids:
+                if f.name in fields or f.field_id.name in fields:
+                    result |= record
```
Comment (nhatnm0612) [inline spp_import_match/models/import_match.py]:
```suggestion
        available = self.search([("model_name", "=", model_name)], limit=1)
        if not available:
            return [], []
        field_to_match = []
        for f in available.field_ids:
            if f.name in fields or f.field_id.name in fields:
                result |= available
            field_to_match.append(f.name)

```

**Comment by emjay0921** [comment]
Added a boolean field in import match to have an option if matched data will overwitten or ignored.

**Review by gonzalesedwin1123 — CHANGES_REQUESTED** [review]
Fix the errors in the automated test

File: spp_import_match/models/import_match.py (hunk 1 above)
Comment (emjay0921) [reply spp_import_match/models/import_match.py]: This is edited, sql_constraints is not good here is we should allow different sub-fields for a one2many field.

File: spp_import_match/models/import_match.py (hunk 2 above)
Comment (emjay0921) [reply spp_import_match/models/import_match.py:35]: This will be removed, as we can have multiple Import Match Configuration for a single model.

**Pull request merged**